### PR TITLE
Refactor datapackage, to work with ckanext-downloadall

### PR DESCRIPTION
**Pull request by davidread**
I've factored out a couple of funcs, for use by ckanext-download: https://github.com/davidread/ckanext-downloadall/pull/3

TODO
* [ ] Revert `populate_datastore_fields` alone after
* [ ] Unit tests for the new funcs

### CODE REVIEW

File: ckanapi/datapackage.py [hunk 1]
```diff
@@ -50,69 +50,93 @@ def create_datapackage(record, base_path, stderr):
...
+
 
-def populate_datastore_fields(ckan, dataset):
+def resource_filename(dres, existing_filenames):
+    # prefer resource names from datapackage metadata
+    name = dres['name']
+    ext = slugify.slugify(dres['format'])
+    if name.endswith(ext):
+        name = name[:-len(ext)]
+    # deduplicate
```
Comment (wardi) [inline ckanapi/datapackage.py]: the datapackage code already deduplicates the names, no?

File: ckanapi/datapackage.py [hunk 2]
```diff
@@ -15,12 +15,12 @@
 }
 
 
-def create_resource(resource, datapackage_dir, stderr):
+def create_resource(resource, filename, datapackage_dir, stderr):
     # Resources can have multiple sources with the same name, or names with
     # filesystem-unfriendly characters, or URLs with a trailing slash, or
     # multiple URLs with the same final path component, so we're just going to
     # name all downloads using the resource's UID.
```
Comment (wardi) [inline ckanapi/datapackage.py]: outdated comment to remove

File: ckanapi/datapackage.py (hunk 1 above)
Comment (wardi) [reply ckanapi/datapackage.py]: https://github.com/ckan/ckanapi/blob/d3a7990230c15d69665990501e89a703bafa283f/ckanapi/datapackage.py#L196-L204

Comment (davidread) [reply ckanapi/datapackage.py]: agreed, I'll remove this

**Comment by davidread** [comment]
I'm going to do a few more tests before this is done - hopefully have something later.

**Comment by wardi** [comment]
looks great. maybe I should make the other tests use pyfakefs too at some point.

**Comment by davidread** [comment]
I see there is a pytest plug-in for pyfakefs, so if we switch to that I
think we don’t need to activate it in the test code.

**Pull request merged**